iroh-net/src/discovery/local_node_discovery.rs

```diff
@@ -0,0 +1,250 @@
+use std::{
+    collections::{BTreeSet, HashMap},
+    net::{IpAddr, SocketAddr},
+    time::Duration,
+};
+
+use anyhow::Result;
+use derive_more::FromStr;
+use futures_lite::{stream::Boxed as BoxStream, StreamExt};
+
+use flume::Sender;
+use iroh_base::key::PublicKey;
+use swarm_discovery::{Discoverer, DropGuard, Peer};
+use tokio::task::{JoinHandle, JoinSet};
+
+use crate::{
+    discovery::{Discovery, DiscoveryItem},
+    AddrInfo, Endpoint, NodeId,
+};
+
+/// The n0 local node discovery name
+// TODO(ramfox): bikeshed
+const N0_MDNS_SWARM: &str = "iroh.local.node.discovery";
+
+/// Provenance string
+// TODO(ramfox): bikeshed
+const PROVENANCE: &str = "local.node.discovery";
+
+/// How long we will wait before we stop sending discovery items
+const DISCOVERY_DURATION: Duration = Duration::from_secs(10);
+
+/// Discovery using `swarm-discovery`, a variation on mdns
+#[derive(Debug)]
+pub struct LocalNodeDiscovery {
+    handle: JoinHandle<()>,
+    sender: Sender<Message>,
+}
+
+impl Drop for LocalNodeDiscovery {
+    fn drop(&mut self) {
+        self.handle.abort();
+    }
+}
+
+#[derive(Debug)]
+enum Message {
+    Discovery((String, Peer)),
+    SendAddrs((NodeId, Sender<Result<DiscoveryItem>>)),
+    ChangeLocalAddrs(AddrInfo),
+    Timeout(NodeId),
+}
+
+impl LocalNodeDiscovery {
+    /// Create a new LocalNodeDiscovery Service.
+    ///
+    /// This starts a `Discoverer` that broadcasts your addresses and receives addresses from other nodes in your local network.
+    pub fn new(node_id: NodeId) -> Self {
+        tracing::debug!("Creating new LocalNodeDiscovery service");
+        let (send, recv) = flume::bounded(64);
+        let task_sender = send.clone();
+        let rt = tokio::runtime::Handle::current();
+        let handle = tokio::spawn(async move {
+            let mut guard = match LocalNodeDiscovery::spawn_discoverer(
+                node_id,
+                task_sender.clone(),
+                BTreeSet::new(),
+                &rt,
+            ) {
+                Ok(guard) => Some(guard),
+                Err(e) => {
+                    tracing::error!("LocalNodeDiscovery error creating discovery service: {e}");
+                    return;
+                }
+            };
+            let mut node_addrs: HashMap<PublicKey, Peer> = HashMap::default();
+            let mut senders: HashMap<PublicKey, Sender<Result<DiscoveryItem>>> = HashMap::default();
+            let mut timeouts = JoinSet::new();
```

Comment:
These seem to be serving conflicting purposes. 

On one hand we store a single `Sender` per node, but this is overwritten (and thus the previous one being dropped) for any new call for the same node. In particular this means that concurrent requests (`Discovery::resolve`) to the same node when no address is known will result in the first one receiving nothing, because the second one effectively cancels it. 

In the same fashion, the timeouts of any previous request take over the timeouts of new requests. So, if we do a request at `t=0` a new timeout task is spawned, then another request at `t=5` the first one will have a timeout of 10secs, the second one of 5 seconds because there is nothing cancelling, or otherwise differentiating, the first timeout.

If we want to reuse senders for all concurrent requests of the same peer, `senders` needs to be moved out of the run loop an be shared between the "service" and "handler". If we want timeouts to be applied per request, these need to be moved to the `Discovery::resolve" impl itself

Reply:
i've added a unique id to each request, so we can store multiple senders for each node id & added a test to make sure that we can handle multiple streams for the same node id.